rpc/notification.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package rpc
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+)
+
+var (
+	NotificationsUnsupportedErr = errors.New("notifications not supported")
+	NotificationNotFoundErr     = errors.New("notification not found")
+	notifierStoppedErr          = errors.New("unable to send notification")
+	notificationQueueFullErr    = errors.New("too many pending notifications")
+)
+
+var unsubSignal struct{}
+
+// UnsubscribeCallback defines a callback that is called when a subcription ends.
+// It receives the subscription id as argument.
+type UnsubscribeCallback func(id string)
+
+// notification is a helper object that holds event data for a subscription
+type notification struct {
+	sub  *bufferedSubscription // subscription id
+	data interface{}           // event data
+}
+
+// A notifier describes the interface for objects that can send create subscriptions
+type Notifier interface {
+	// Create a new subscription. The given callback is called when this subscription
+	// is cancelled (e.g. client send an unsubscribe, connection closed).
+	NewSubscription(UnsubscribeCallback) (Subscription, error)
+	// Cancel subscription
+	Unsubscribe(id string) error
+}
+
+// Subscription defines the interface for objects that can notify subscribers
+type Subscription interface {
+	// Inform client of an event
+	Notify(data interface{}) error
+	// Unique identifier
+	Id() string
+	// Cancel subscription
+	Cancel() error
+}
+
+// bufferedSubscription is a subscription that uses a bufferedNotifier to send
+// notifications to subscribers.
+type bufferedSubscription struct {
+	id               string
+	once             sync.Once
```

Comment:
Please name this something else. E.g. `closer` will have a meaningful name. Otherwise `once` doesn't really convey what it does.